File: backend/app/ai/reasoning_controls.py
"""文件功能：规范化 Models.dev 推理控制元数据，并计算协议可安全发送的控制能力。"""

from __future__ import annotations

from typing import Any

REASONING_CONTROL_TYPES = {"toggle", "effort", "budget_tokens"}

PROTOCOL_REASONING_CONTROLS: dict[str, frozenset[str]] = {
    "openai_chat": frozenset({"toggle", "effort"}),
    "openai_compatible_chat": frozenset({"effort"}),
    "openrouter_chat": frozenset({"toggle", "effort"}),
    "google_chat": frozenset({"effort"}),
    "alibaba_openai_compatible": frozenset({"toggle", "budget_tokens"}),
    "deepseek_openai_compatible": frozenset({"toggle", "effort"}),
    "xiaomi_openai_compatible": frozenset({"toggle"}),
    "ollama_openai_compatible": frozenset({"toggle", "effort"}),
}


def normalize_reasoning_options(value: Any) -> dict[str, Any]:
    """把 Models.dev 当前的选项对象数组和旧格式统一为稳定的内部结构。"""

    controls: list[dict[str, Any]] = []
    if isinstance(value, dict) and isinstance(value.get("type"), str):
        controls = [value]
    elif isinstance(value, (list, tuple)):
        if all(isinstance(item, str) for item in value):
            controls = [{"type": "effort", "values": list(value)}]
        else:
            controls = [item for item in value if isinstance(item, dict)]
    elif isinstance(value, dict):
        nested = value.get("options")
        if isinstance(nested, (list, tuple)) and any(isinstance(item, dict) for item in nested):
            controls = [item for item in nested if isinstance(item, dict)]
        else:
            effort = value.get("effort") or value.get("efforts") or value.get("levels") or nested
            if isinstance(effort, dict):
                effort = list(effort)
            if isinstance(effort, (list, tuple)):
                controls.append({"type": "effort", "values": list(effort)})
            if isinstance(value.get("budget_tokens"), dict):
                controls.append({"type": "budget_tokens", **value["budget_tokens"]})
            for control_type in value.get("types", []) if isinstance(value.get("types"), list) else []:
                if control_type == "toggle":
                    controls.append({"type": "toggle"})

    types: list[str] = []
    efforts: list[str] = []
    budget: dict[str, int] = {}
    for control in controls:
        control_type = str(control.get("type") or "").strip()
        if control_type not in REASONING_CONTROL_TYPES:
            continue
        if control_type not in types:
            types.append(control_type)
        if control_type == "effort":
            values = control.get("values")
            if isinstance(values, (list, tuple)):
                for item in values:
                    normalized = str(item).strip() if isinstance(item, str) else ""
                    if normalized and normalized not in efforts:
                        efforts.append(normalized)
        if control_type == "budget_tokens":
            for key in ("min", "max"):
                item = control.get(key)
                if isinstance(item, int) and item > 0:
                    budget[key] = item

    result: dict[str, Any] = {"types": types}
    if efforts:
        result["effort"] = efforts
    if budget:
        result["budget_tokens"] = budget
    return result


def reasoning_control_types(value: Any) -> frozenset[str]:
    """返回模型目录明确声明的推理控制类型。"""

    return frozenset(normalize_reasoning_options(value).get("types", []))


def reasoning_effort_options(value: Any) -> list[str]:
    """返回模型明确公布的推理强度枚举。"""

    options = normalize_reasoning_options(value).get("effort", [])
    return [item for item in options if isinstance(item, str)]


def reasoning_budget_limits(value: Any) -> dict[str, int]:
    """返回模型公布的推理 token 预算边界。"""

    limits = normalize_reasoning_options(value).get("budget_tokens", {})
    return {key: item for key, item in limits.items() if key in {"min", "max"} and isinstance(item, int)}


def protocol_reasoning_controls(protocol_key: str) -> frozenset[str]:
    """返回固定协议转换器已经实现且可安全发送的推理控制类型。"""

    return PROTOCOL_REASONING_CONTROLS.get(protocol_key, frozenset())


def supports_explicit_disable(options: Any, protocol_key: str) -> bool:
    """仅当模型声明与协议转换器均能表达关闭语义时返回真。"""

    model_controls = reasoning_control_types(options)
    protocol_controls = protocol_reasoning_controls(protocol_key)
    return bool(
        "toggle" in model_controls.intersection(protocol_controls)
        or ("effort" in protocol_controls and "none" in reasoning_effort_options(options))
    )

File: editor/src/utils/reasoning-controls.ts
/**
 * 文件功能：解析 Models.dev 推理选项，并计算模型元数据与固定协议转换器的能力交集。
 */

/** 兼容当前选项对象数组、同步后的规范结构和旧缓存结构。 */
export function reasoningOptionRecords(raw: unknown): Record<string, unknown>[] {
  if (Array.isArray(raw)) return raw.filter(isOptionRecord)
  if (!raw || typeof raw !== 'object') return []
  const record = raw as Record<string, unknown>
  if (typeof record.type === 'string') return [record]
  if (Array.isArray(record.options) && record.options.some(isOptionRecord)) return record.options.filter(isOptionRecord)
  return []
}

/** 提取模型声明的 toggle、effort 和 budget_tokens 控制类型。 */
export function reasoningControls(raw: unknown): Set<string> {
  const controls = new Set(reasoningOptionRecords(raw).map(item => String(item.type ?? '')).filter(Boolean))
  if (!raw || typeof raw !== 'object' || Array.isArray(raw)) return controls
  const record = raw as Record<string, unknown>
  if (Array.isArray(record.types)) record.types.forEach(item => controls.add(String(item)))
  if (record.effort || record.efforts || record.levels) controls.add('effort')
  if (record.budget_tokens) controls.add('budget_tokens')
  return controls
}

/** 提取目录公布的 effort 枚举，包含 none 以便判断是否能够关闭。 */
export function reasoningEfforts(raw: unknown): string[] {
  const values: unknown[] = []
  for (const record of reasoningOptionRecords(raw)) {
    if (record.type === 'effort' && Array.isArray(record.values)) values.push(...record.values)
  }
  if (raw && typeof raw === 'object' && !Array.isArray(raw)) {
    const record = raw as Record<string, unknown>
    const legacy = record.effort ?? record.efforts ?? record.levels
    if (Array.isArray(legacy)) values.push(...legacy)
    else if (legacy && typeof legacy === 'object') values.push(...Object.keys(legacy))
    if (Array.isArray(record.options) && record.options.every(item => typeof item === 'string')) values.push(...record.options)
  }
  return [...new Set(values.filter((item): item is string => typeof item === 'string' && Boolean(item.trim())).map(item => item.trim()))]
}

/** 提取目录公布的推理 token 预算上下界。 */
export function reasoningBudgetLimits(raw: unknown): { min?: number; max?: number } {
  if (!raw || typeof raw !== 'object' || Array.isArray(raw)) {
    const record = reasoningOptionRecords(raw).find(item => item.type === 'budget_tokens')
    return positiveLimits(record)
  }
  const normalized = raw as Record<string, unknown>
  if (normalized.budget_tokens && typeof normalized.budget_tokens === 'object') {
    return positiveLimits(normalized.budget_tokens as Record<string, unknown>)
  }
  return positiveLimits(reasoningOptionRecords(raw).find(item => item.type === 'budget_tokens'))
}

/** 返回当前固定协议转换器真正实现的推理控制能力。 */
export function protocolReasoningControls(protocol: string): Set<string> {
  const controls: Record<string, string[]> = {
    openai_chat: ['toggle', 'effort'],
    openai_compatible_chat: ['effort'],
    openrouter_chat: ['toggle', 'effort'],
    google_chat: ['effort'],
    alibaba_openai_compatible: ['toggle', 'budget_tokens'],
    deepseek_openai_compatible: ['toggle', 'effort'],
    xiaomi_openai_compatible: ['toggle'],
    ollama_openai_compatible: ['toggle', 'effort'],
  }
  return new Set(controls[protocol] ?? [])
}

/** 判断模型声明与固定协议是否共同支持明确关闭推理。 */
export function canDisableReasoning(raw: unknown, protocol: string): boolean {
  const modelControls = reasoningControls(raw)
  const protocolControls = protocolReasoningControls(protocol)
  return modelControls.has('toggle') && protocolControls.has('toggle')
    || protocolControls.has('effort') && reasoningEfforts(raw).includes('none')
}

/** 计算一次 Run 可选的推理模式，auto 永远作为安全回退。 */
export function availableReasoningModes(
  supportsReasoning: boolean,
  raw: unknown,
  protocol: string,
): Array<'auto' | 'disabled' | 'effort' | 'budget_tokens'> {
  const modes: Array<'auto' | 'disabled' | 'effort' | 'budget_tokens'> = ['auto']
  if (!supportsReasoning) return modes
  const modelControls = reasoningControls(raw)
  const protocolControls = protocolReasoningControls(protocol)
  if (canDisableReasoning(raw, protocol)) modes.push('disabled')
  if (modelControls.has('effort') && protocolControls.has('effort') && reasoningEfforts(raw).some(value => value !== 'none')) modes.push('effort')
  if (modelControls.has('budget_tokens') && protocolControls.has('budget_tokens')) modes.push('budget_tokens')
  return modes
}

/** 缩小 unknown 的目录选项对象类型。 */
function isOptionRecord(item: unknown): item is Record<string, unknown> {
  return Boolean(item) && typeof item === 'object' && !Array.isArray(item)
}

/** 只接受正整数预算边界，忽略目录中的无效扩展值。 */
function positiveLimits(record: Record<string, unknown> | undefined): { min?: number; max?: number } {
  const result: { min?: number; max?: number } = {}
  if (Number.isInteger(record?.min) && Number(record?.min) > 0) result.min = Number(record?.min)
  if (Number.isInteger(record?.max) && Number(record?.max) > 0) result.max = Number(record?.max)
  return result
}
